### PR TITLE
Add support of PHP8 static return type for withers

### PR DESCRIPTION
**Pull request by l-vo**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | https://github.com/orgs/symfony/projects/1#card-35852557
| License       | MIT
| Doc PR        | symfony/symfony-docs#13619

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php [hunk 1]
```diff
@@ -51,7 +51,10 @@ protected function processValue($value, bool $isRoot = false)
             while (true) {
                 if (false !== $doc = $r->getDocComment()) {
                     if (false !== stripos($doc, '@required') && preg_match('#(?:^/\*\*|\n\s*+\*)\s*+@required(?:\s|\*/$)#i', $doc)) {
-                        if (preg_match('#(?:^/\*\*|\n\s*+\*)\s*+@return\s++static[\s\*]#i', $doc)) {
+                        $reflectionType = $reflectionMethod->getReturnType();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:54]: only is hasReturnType

File: src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php [hunk 2]
```diff
@@ -51,7 +51,10 @@ protected function processValue($value, bool $isRoot = false)
...
+                        if ($reflectionType instanceof \ReflectionNamedType && 'static' == $reflectionType->getName()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:55]: `===`

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowireRequiredMethodsPassTest.php [hunk 3]
```diff
@@ -99,4 +99,28 @@ public function testWitherInjection()
         ];
         $this->assertSame($expected, $methodCalls);
     }
+
+    public function testWitherWithStaticReturnTypeInjection()
+    {
+        if (version_compare(\PHP_VERSION_ID, 80000, 'lt')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowireRequiredMethodsPassTest.php:105]: `@requires PHP 8` on the method

File: src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php [hunk 4]
```diff
@@ -1624,6 +1625,26 @@ public function testWither()
         $this->assertInstanceOf(Foo::class, $wither->foo);
     }
 
+    public function testWitherWithStaticReturnType()
+    {
+        if (version_compare(\PHP_VERSION_ID, 80000, 'lt')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php:1630]: annotation also

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 5]
```diff
@@ -1362,6 +1363,32 @@ public function testWither()
         $this->assertInstanceOf(Foo::class, $wither->foo);
     }
 
+    public function testWitherWithStaticReturnType()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php:1366]: same

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php [hunk 6]
```diff
@@ -240,7 +244,8 @@ public function setFoo(Foo $foo)
         // should be called
     }
 
-    /** @inheritdoc*/ // <- brackets are missing on purpose
+    /** {@inheritdoc}*/ // <- brackets are missing on purpose
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php:247]: please read the comment :)

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php [hunk 7]
```diff
@@ -277,6 +282,7 @@ public function setFoo(Foo $foo)
 
     /**
      * @required
+     *
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php:285]: better keep a is (same below)

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes_8.php [hunk 8]
```diff
@@ -0,0 +1,18 @@
+<?php
+namespace Symfony\Component\DependencyInjection\Tests\Compiler;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes_8.php:2]: missing blank line

**Review by nicolas-grekas — COMMENTED** [review]
Actually, I just realized the check should also verify that `@return $this` is NOT present.
We really miss the `$this` return type, which would provide much better clarity and would save us many annotations.
With the current state, we'll still have to use docblocks to help differentiate those two situations.

/cc @nikic FYI, in case this can be considered for PHP v8

File: src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php (hunk 1 above)
Comment (l-vo) [reply src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php]: Indeed, my bad !

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php (hunk 6 above)
Comment (l-vo) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php]: Arf bad idea too run cs-fixer on these files :/

**Comment by l-vo** [comment]
Updated 🙂

**Review by nicolas-grekas — APPROVED** [review]

**Comment by l-vo** [comment]
I still have a problem on travis 7.2/php_extra=7.4, my check on PHP_VERSION_ID seems not working on this environment 🤔 Do I miss something specific to this environment ?

**Comment by nicolas-grekas** [comment]
You need to list the new file in `.github/patch-types.php`

File: .github/patch-types.php [hunk 9]
```diff
@@ -30,6 +30,7 @@
         case false !== strpos($file, '/src/Symfony/Component/Serializer/Tests/Normalizer/Features/ObjectOuter.php'):
         case false !== strpos($file, '/src/Symfony/Component/VarDumper/Tests/Fixtures/NotLoadableClass.php'):
         case false !== strpos($file, '/src/Symfony/Component/VarDumper/Tests/Fixtures/Php74.php') && \PHP_VERSION_ID < 70400:
+        case false !== strpos($file, '/src/Symfony/Component/DependencyInjection/Tests/Fixtures/WitherStaticReturnType.php'):
```
Comment (nicolas-grekas) [inline .github/patch-types.php:33]: Alpha order :)

**Comment by l-vo** [comment]
Updated 🙂

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @l-vo.

**Pull request merged**